<a href="https://colab.research.google.com/github/vanesamp07/Python/blob/main/PruebasdeHipotesis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
pip install numpy scipy

Note: you may need to restart the kernel to use updated packages.


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as st
import statsmodels.api as sm

In [ ]:
df = pd.read_csv('earnings.csv')

In [ ]:
df.head()

height  weight  male     earn  earnk ethnicity  education  \
0      74   210.0     1  50000.0   50.0     White       16.0   
1      66   125.0     0  60000.0   60.0     White       16.0   
2      64   126.0     0  30000.0   30.0     White       16.0   
3      65   200.0     0  25000.0   25.0     White       17.0   
4      63   110.0     0  50000.0   50.0     Other       16.0   

   mother_education  father_education  walk  exercise  smokenow  tense  angry  \
0              16.0              16.0     3         3       2.0    0.0    0.0   
1              16.0              16.0     6         5       1.0    0.0    0.0   
2              16.0              16.0     8         1       2.0    1.0    1.0   
3              17.0               NaN     8         1       2.0    0.0    0.0   
4              16.0              16.0     5         6       2.0    0.0    0.0   

   age  
0   45  
1   58  
2   29  
3   57  
4   91

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1816 entries, 0 to 1815
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   height            1816 non-null   int64  
 1   weight            1789 non-null   float64
 2   male              1816 non-null   int64  
 3   earn              1816 non-null   float64
 4   earnk             1816 non-null   float64
 5   ethnicity         1816 non-null   object 
 6   education         1814 non-null   float64
 7   mother_education  1572 non-null   float64
 8   father_education  1521 non-null   float64
 9   walk              1816 non-null   int64  
 10  exercise          1816 non-null   int64  
 11  smokenow          1815 non-null   float64
 12  tense             1815 non-null   float64
 13  angry             1815 non-null   float64
 14  age               1816 non-null   int64  
dtypes: float64(9), int64(5), object(1)
memory usage: 212.9+ KB


In [ ]:
Desviacion_estandar_earn = np.std(df["earn"])
Desviacion_estandar_earn

22525.56058921212

In [ ]:
Desviacion_estandar_height = np.std(df["height"])
Desviacion_estandar_height

3.830767219735599

In [ ]:
Desviacion_estandar_age = np.std(df["age"])
Desviacion_estandar_age

17.15694594337249

In [ ]:
#Realizamos 12 pruebas con muestras de tamaño n=25
#Los niveles de significancia serán menor a (alfa=0.1)
#Si p-value es menor a alfa se rechaza la Hipotesis nula (Ho) es favor de la Hipotesis alternativa (H1)
#Si p-value es mayor a la alfa, se acepta la Ho

n = 25

np.random.seed(0)
earn_prueba1 = np.random.normal(20000, Desviacion_estandar_earn, n)
earn_prueba2 = np.random.normal(35000, Desviacion_estandar_earn, n)
earn_prueba3 = np.random.normal(25000, Desviacion_estandar_earn, n)
earn_prueba4 = np.random.normal(32000, Desviacion_estandar_earn, n)

#earn, prueba 1
#Alfa = 0.05
Nivel_confianza_1 = 95 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["earn"], 20000, alternative='two-sided')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.05:
    print (f"Prueba 1: con nivel de confianza del {Nivel_confianza_1}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 1: con nivel de confianza del {Nivel_confianza_1}%, se rechaza la Ho en favor de la H1")

#earn, prueba 2
#Alfa = 0.01
Nivel_confianza_2 = 99 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["earn"], 35000, alternative='two-sided')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.01:
    print (f"Prueba 2: con nivel de confianza del {Nivel_confianza_2}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 2: con nivel de confianza del {Nivel_confianza_2}%, se rechaza la Ho en favor de la H1")

#earn, prueba 3
#Alfa = 0.07
Nivel_confianza_3 = 93 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["earn"], 25000, alternative='greater')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.07:
   print (f"Prueba 3: con nivel de confianza del {Nivel_confianza_3}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 3: con nivel de confianza del {Nivel_confianza_3}%, se rechaza la Ho en favor de la H1")

#earn, prueba 4
#Alfa = 0.03
Nivel_confianza_4 = 97 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["earn"], 32000, alternative='less')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.03:
     print (f"Prueba 4: con nivel de confianza del {Nivel_confianza_4}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 4: con nivel de confianza del {Nivel_confianza_4}%, se rechaza la Ho en favor de la H1")

Valor t: 2.169892820407049
Valor p: 0.03014431570582848

Prueba 1: con nivel de confianza del 95%, se rechaza la Ho en favor de la H1
Valor t: -26.199756387358775
Valor p: 1.3416968508390565e-128

Prueba 2: con nivel de confianza del 99%, se rechaza la Ho en favor de la H1
Valor t: -7.286656915514892
Valor p: 0.9999999999997642

Prueba 3: con nivel de confianza del 93%, se acepta la hipótesis nula
Valor t: -20.52582654580561
Valor p: 1.1565856821380943e-84

Prueba 4: con nivel de confianza del 97%, se rechaza la Ho en favor de la H1


In [ ]:
n = 25

np.random.seed(0)
height_prueba1 = np.random.normal(55, Desviacion_estandar_height, n)
height_prueba2 = np.random.normal(68, Desviacion_estandar_height, n)
height_prueba3 = np.random.normal(63, Desviacion_estandar_height, n)
height_prueba4 = np.random.normal(70, Desviacion_estandar_height, n)

#height, prueba 5
#Alfa = 0.05
Nivel_confianza_5 = 95 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["height"], 55, alternative='two-sided')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.05:
    print (f"Prueba 5: con nivel de confianza del {Nivel_confianza_5}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 5: con nivel de confianza del {Nivel_confianza_5}%, se rechaza la Ho en favor de la H1")

#height, prueba 6
#Alfa = 0.01
Nivel_confianza_6 = 99 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["height"], 68, alternative='two-sided')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.01:
    print (f"Prueba 6: con nivel de confianza del {Nivel_confianza_6}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 6: con nivel de confianza del {Nivel_confianza_6}%, se rechaza la Ho en favor de la H1")

#height, prueba 7
#Alfa = 0.07
Nivel_confianza_7 = 93 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["height"], 63, alternative='greater')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.07:
    print (f"Prueba 7: con nivel de confianza del {Nivel_confianza_7}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 7: con nivel de confianza del {Nivel_confianza_7}%, se rechaza la Ho en favor de la H1")

#height, prueba 8
#Alfa = 0.03
Nivel_confianza_8 = 97 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["height"], 70, alternative='less')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.03:
    print (f"Prueba 8: con nivel de confianza del {Nivel_confianza_8}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 8: con nivel de confianza del {Nivel_confianza_8}%, se rechaza la Ho en favor de la H1")

Valor t: 128.65956807993143
Valor p: 0.0

Prueba 5: con nivel de confianza del 95%, se rechaza la Ho en favor de la H1
Valor t: -15.916331926304927
Valor p: 1.7073763640776962e-53

Prueba 6: con nivel de confianza del 99%, se rechaza la Ho en favor de la H1
Valor t: 39.68978346070906
Valor p: 7.576455018925499e-249

Prueba 7: con nivel de confianza del 93%, se rechaza la Ho en favor de la H1
Valor t: -38.15877808111052
Valor p: 9.85569780780098e-235

Prueba 8: con nivel de confianza del 97%, se rechaza la Ho en favor de la H1


In [ ]:
n = 25

np.random.seed(0)
age_prueba1 = np.random.normal(44, Desviacion_estandar_age, n)
age_prueba2 = np.random.normal(56, Desviacion_estandar_age, n)
age_prueba3 = np.random.normal(51, Desviacion_estandar_age, n)
age_prueba4 = np.random.normal(62, Desviacion_estandar_age, n)

#age, prueba 9
#Alfa = 0.05
Nivel_confianza_9 = 95 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["age"], 44, alternative='two-sided')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.05:
    print (f"Prueba 9: con nivel de confianza del {Nivel_confianza_9}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 9: con nivel de confianza del {Nivel_confianza_9}%, se rechaza la Ho en favor de la H1")

#age, prueba 10
#Alfa = 0.01
Nivel_confianza_10 = 99 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["age"], 56, alternative='two-sided')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.01:
    print (f"Prueba 10: con nivel de confianza del {Nivel_confianza_10}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 10: con nivel de confianza del {Nivel_confianza_10}%, se rechaza la Ho en favor de la H1")

#age, prueba 11
#Alfa = 0.07
Nivel_confianza_11 = 93 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["age"], 51, alternative='greater')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.07:
    print (f"Prueba 11: con nivel de confianza del {Nivel_confianza_11}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 11: con nivel de confianza del {Nivel_confianza_11}%, se rechaza la Ho en favor de la H1")


#age, prueba 12
#Alfa = 0.03
Nivel_confianza_12 = 97 #expresado en porcentaje
t_stat, p_value = st.ttest_1samp(df["age"], 62, alternative='less')
print(f"Valor t: {t_stat}")
print(f"Valor p: {p_value}\n")
if p_value > 0.03:
     print (f"Prueba 12: con nivel de confianza del {Nivel_confianza_12}%, se acepta la hipótesis nula")
else:
    print (f"Prueba 12: con nivel de confianza del {Nivel_confianza_12}%, se rechaza la Ho en favor de la H1")


Valor t: -2.6458392632055547
Valor p: 0.008219217702944792

Prueba 9: con nivel de confianza del 95%, se rechaza la Ho en favor de la H1
Valor t: -32.44332206619021
Valor p: 1.6839926789785557e-182

Prueba 10: con nivel de confianza del 99%, se rechaza la Ho en favor de la H1
Valor t: -20.02770423161327
Valor p: 1.0

Prueba 11: con nivel de confianza del 93%, se acepta la hipótesis nula
Valor t: -47.342063467682536
Valor p: 1.43007e-319

Prueba 12: con nivel de confianza del 97%, se rechaza la Ho en favor de la H1


In [ ]:
#Calcular media poblacion para cada variable
Mediapoblacion_earn = np.mean(df["earn"])
Mediapoblacion_height = np.mean(df["height"])
Mediapoblacion_age = np.mean(df["age"])

print(f"La media poblacion de la variable earn es igual a {Mediapoblacion_earn}")
print(f"La media poblacion de la variable height es igual a {Mediapoblacion_height}")
print(f"La media poblacion de la variable age es igual a {Mediapoblacion_age}")

La media poblacion de la variable earn es igual a 21147.296255506608
La media poblacion de la variable height es igual a 66.56883259911895
La media poblacion de la variable age es igual a 42.934471365638764


In [ ]:
#Vamos a coomparar los resultados con la variable earn
#Debemos tener en cuenta que para earn se acepto solamente la Ho numero 3 con un nivel de confiabilidad del 93%

Diferencia1 = Mediapoblacion_earn - 20000
Diferencia2 = Mediapoblacion_earn - 35000
Diferencia3 = Mediapoblacion_earn - 25000
Diferencia4 = Mediapoblacion_earn - 32000

Diferencia1

1147.2962555066078

In [ ]:
Diferencia2

-13852.703744493392

In [ ]:
Diferencia3

-3852.703744493392

In [ ]:
Diferencia4

-10852.703744493392

In [ ]:
#Imprimimos nuevamente la STD como base de la coomparación
Desviacion_estandar_earn

22525.56058921212

Caso prueba 1= efectivamente es correcto el resultado, se rechazo la Ho en favor de la H1.
Es adecuado el resultado porque la media obtenida es diferente a la Ho

Caso prueba 2= efectivamente es correcto el resultado, se rechazo la Ho en favor de la H1.
Es adecuado el resultado porque la media obtenida es diferente a la Ho

Caso prueba 3= no es correcto el resultado, no se debio aceptar la Ho porque la media no es igual a 25.000

Caso prueba 2= efectivamente es correcto el resultado, se rechazo la Ho en favor de la H1.
Es adecuado el resultado porque la media obtenida es menor a la Ho

Pese a lo anterior, teniendo en cuenta que la desviacion es igual a +- 22525, las diferencias entre las medias hipoteticas y la media real se encuentran dentro de los rangos

In [ ]:
#Vamos a coomparar los resultados con la variable height
#Debemos tener en cuenta que para height, se rechazaron las cuatro Ho

Diferencia1 = Mediapoblacion_height - 55
Diferencia2 = Mediapoblacion_height - 68
Diferencia3 = Mediapoblacion_height - 63
Diferencia4 = Mediapoblacion_height - 70

Diferencia1

11.568832599118949

In [ ]:
Diferencia2

-1.4311674008810513

In [ ]:
Diferencia3

3.5688325991189487

In [ ]:
Diferencia4

-3.4311674008810513

In [ ]:
#Imprimimos nuevamente la STD como base de la coomparación
Desviacion_estandar_height

3.830767219735599

Para los cuatro casos, esta bien que se haya rechazado la Ho en favor de la alternativa.
Para el caso 3, efectivamente la media es mayor a 63 como lo establece la H1.
Para el caso 4, efectivamente la media es menor a 70 como lo establece la H1.
Ninguna de las cuatro Ho establecen que la media es igual al resultado obtenido (66.56).
Sin embargo, la std es +- 3.83, por lo que para el caso 2, el caso 3 y el caso 4, se encuentra dentro de los intervalos.

In [ ]:
#Vamos a coomparar los resultados con la variable age
#Debemos tener en cuenta que para earn se acepto solamente la Ho numero 3 con un nivel de confiabilidad del 93%

Diferencia1 = Mediapoblacion_age - 44
Diferencia2 = Mediapoblacion_age - 56
Diferencia3 = Mediapoblacion_age - 51
Diferencia4 = Mediapoblacion_age - 62

Diferencia1

-1.065528634361236

In [ ]:
Diferencia2

-13.065528634361236

In [ ]:
Diferencia3

-8.065528634361236

In [ ]:
Diferencia4

-19.065528634361236

In [ ]:
#Imprimimos nuevamente la STD como base de la coomparación
Desviacion_estandar_age

17.15694594337249

Para el caso 1, 2 y 4, esta bien que se haya rechazado la Ho en favor de la alternativa.
Para el caso 3, se debio rechazar la Ho ya que la media obtenida equivale a 42.93 y la Ho dice que la media es igual a 51.
Para el caso 3, la H1 dice que la media es mayor a 51, sin embargo, tampoco es cierto.
Para el caso 4, la H1 nos dice que la media es menor a 62, lo cual es correcto.
Sin embargo, la std es +- 17.15, por lo que para los 4 casos, se encuentra dentro de los intervalos.

In [ ]:
#Contar la cantidad de hombres en en el dataset
num_hombres = df[df["male"] == "1"].shape[0]

#Proporción muestral de hombres en la muestra
proporcion_muestral = num_hombres / n

#Nivel de confianza
confianza = 0.95

intervalos = {}

#Calcular el intervalo de confianza utilizando statsmodels
intervalo_confianza = sm.stats.proportion_confint(num_hombres, n, alpha=1 - confianza)
intervalos[f'{int(confianza * 100)}%'] = intervalo_confianza

#Imprimir el resultado
print(f"Intervalo de confianza del {confianza * 100}% para la proporción de hombres en el dataset considerando una muestra de 50 individuos es: ")
print(f"({intervalo_confianza[0]}, {intervalo_confianza[1]})")

Intervalo de confianza del 95.0% para la proporción de hombres en el dataset considerando una muestra de 50 individuos es: 
(0.0, 0.0)


- Considerando el dataset como una muestra, verifica si el género (male=1 significa hombre) inuye sobre el sueldo earn de las personas.

Ho = No hay diferencia en el sueldo entre hombres y mujeres

H1 = Si hay diferencia en el sueldo entre hombres y mujeres

In [ ]:
#Verificar si el genero  influye sobre el sueldo de las personas

#Se divide el dataset por genero
hombres = df[df["male"] == 1]["earn"]
mujeres = df[df["male"] == 0]["earn"]

#Se realiza t-student para muestras independientes
t_st, p_value = st.ttest_ind(hombres, mujeres)
print(f"Valor t= {t_st}")
print(f"Valor p= {p_value}")

#Establecemos un nivel de significancia
nivel_significancia = 0.05

#Establecemos si se aceptará o no la Ho
if p_value < nivel_significancia:
    print(f"Se rechaza la Ho en favor de la H1 alternativa")
else:
    print(f"No hay suficiente evidencia para rechazar la Ho")

Valor t= 13.68318837519989
Valor p= 1.2371946179054945e-40
Se rechaza la Ho en favor de la H1 alternativa


De acuerdo a lo anterior y teniendo en cuenta que el valor p es menor al nivel de significancia que equivale a 0.05, entonces se rechaza la Ho de que no hay diferencia en el sueldo entre hombres y mujeres, por ende se acepta la H1 que establece que sí hay diferencia.